### ISSUE TITLE
Leading minor not positive definite

### ISSUE THREAD

**Issue by iaroslav-ai** [scikit-optimize/scikit-optimize#324]
When I run the following code in python 2.7: 
```
import numpy as np
from skopt import space, Optimizer
from skopt.learning import GaussianProcessRegressor
from skopt.learning.gaussian_process.kernels import Matern

estimator = GaussianProcessRegressor(kernel=Matern())
solver = Optimizer([space.Real(-1.0, 1.0) for _ in range(10)], estimator, acq_optimizer='lbfgs')

for i in range(31):
    p = solver.ask()
    solver.tell(p, np.mean(np.sin(p)))
```
I either get the error like this:
```
Traceback (most recent call last):
  File "/home/iaroslav/scikit-optimize/test_lbfgs.py", line 11, in <module>
    solver.tell(p, np.mean(np.sin(p)))
  File "/home/iaroslav/scikit-optimize/skopt/optimizer/optimizer.py", line 258, in tell
    est.fit(self.space.transform(self.Xi), self.yi)
  File "/home/iaroslav/scikit-optimize/skopt/learning/gaussian_process/gpr.py", line 190, in fit
    super(GaussianProcessRegressor, self).fit(X, y)
  File "/usr/local/lib/python2.7/dist-packages/sklearn/gaussian_process/gpr.py", line 231, in fit
    self.L_ = cholesky(K, lower=True)  # Line 2
  File "/usr/local/lib/python2.7/dist-packages/scipy/linalg/decomp_cholesky.py", line 81, in cholesky
    check_finite=check_finite)
  File "/usr/local/lib/python2.7/dist-packages/scipy/linalg/decomp_cholesky.py", line 30, in _cholesky
    raise LinAlgError("%d-th leading minor not positive definite" % info)
numpy.linalg.linalg.LinAlgError: 19-th leading minor not positive definite
```
or / and I get a warning like this:
```
/home/iaroslav/scikit-optimize/skopt/optimizer/optimizer.py:196: UserWarning: The objective has been evaluated at this point before.
  warnings.warn("The objective has been evaluated "
```
If I set `acq_optimizer='sampling'`, everything works smoothly. It seems as if with the new points selected in lbfgs mode somehow break computation of inverse of kernel matrix (maybe same point or very close points evaluated multiple times?).

**Comment by betatim** [comment]
Maybe @MechCoder can have a look. Sounds like the optimizer quickly finds the minimum and then keeps suggesting the same point. Could you make a plot of what things look like?

This might be a reason to add a "early stopping" feature or some other way for the optimizer to signal "I think we converged on the optimum"

**Comment by iaroslav-ai** [comment]
@betatim indeed it seems that this is what is happening. I modified the code a bit to output best objective found at particular iteration, and for point suggested at this iteration it outputs the minimal L2 distance to all the points that has been already evaluated: 
```
import numpy as np
from skopt import space, Optimizer
from skopt.learning import GaussianProcessRegressor
from skopt.learning.gaussian_process.kernels import Matern

BOUND = 1.6

estimator = GaussianProcessRegressor(kernel=Matern())
solver = Optimizer([space.Real(-BOUND, BOUND) for _ in range(10)], estimator, acq_optimizer='lbfgs')

min_obj = 1.0
print('min. obj. = ',np.mean(np.sin([-min(BOUND, 3.1415926535/2.0)]*10)))

points = []

for i in range(512):
    p = solver.ask()
    obj=np.mean(np.sin(p))
    min_obj=min(obj,min_obj)
    if len(points) > 0:
        print('obj: ', min_obj, 'min. dist: ' ,np.min( np.sum(np.subtract(points,p) ** 2, axis=1)))
    points.append(p)
    solver.tell(p, obj)
```
The output is as follows:
```
('obj: ', -0.087154837008484337, 'min. dist: ', 15.903809035497371)
('obj: ', -0.16131073840519813, 'min. dist: ', 6.2904030923030732)
('obj: ', -0.16131073840519813, 'min. dist: ', 13.733967418136153)
('obj: ', -0.33914765118736023, 'min. dist: ', 7.3950084148436446)
('obj: ', -0.33914765118736023, 'min. dist: ', 11.813180243138442)
('obj: ', -0.33914765118736023, 'min. dist: ', 15.259444034594146)
('obj: ', -0.33914765118736023, 'min. dist: ', 9.9374841812970161)
('obj: ', -0.33914765118736023, 'min. dist: ', 6.8931423072738136)
('obj: ', -0.33914765118736023, 'min. dist: ', 7.6412189998439395)
('obj: ', -0.38788408698688326, 'min. dist: ', 0.16773643031063626)
('obj: ', -0.99957360304150511, 'min. dist: ', 18.701419315995675)
('obj: ', -0.99957360304150511, 'min. dist: ', 10.240000000000002)
('obj: ', -0.99957360304150511, 'min. dist: ', 7.5871300406503623)
('obj: ', -0.99957360304150511, 'min. dist: ', 10.075177984878895)
('obj: ', -0.99957360304150511, 'min. dist: ', 9.6954686979120464)
('obj: ', -0.99957360304150511, 'min. dist: ', 2.06869021168824)
('obj: ', -0.99957360304150511, 'min. dist: ', 14.92577458293929)
('obj: ', -0.99957360304150511, 'min. dist: ', 10.084081182996968)
('obj: ', -0.99957360304150511, 'min. dist: ', 8.6674811860499101)
('obj: ', -0.99957360304150511, 'min. dist: ', 1.9447821702375503)
('obj: ', -0.99957360304150511, 'min. dist: ', 1.883964707153686)
('obj: ', -0.99957360304150511, 'min. dist: ', 2.6575476018273201)
('obj: ', -0.99957360304150511, 'min. dist: ', 3.6447261718997828)
('obj: ', -0.99957360304150511, 'min. dist: ', 1.6450409669675869)
('obj: ', -0.99957360304150511, 'min. dist: ', 1.3185038627548786)
('obj: ', -0.99957360304150511, 'min. dist: ', 0.29516201856265883)
('obj: ', -0.99957360304150511, 'min. dist: ', 0.31583446386638059)
('obj: ', -0.99957360304150511, 'min. dist: ', 0.27419559126155585)
('obj: ', -0.99957360304150511, 'min. dist: ', 0.010473642439863652)
/home/iaroslav/scikit-optimize/skopt/optimizer/optimizer.py:196: UserWarning: The objective has been evaluated at this point before.
  warnings.warn("The objective has been evaluated "
('obj: ', -0.99957360304150511, 'min. dist: ', 0.0)
('obj: ', -0.99957360304150511, 'min. dist: ', 0.23296803183171966)
Traceback (most recent call last):
  File "tst.py", line 23, in <module>
    solver.tell(p, obj)
  File "/home/iaroslav/scikit-optimize/skopt/optimizer/optimizer.py", line 258, in tell
    est.fit(self.space.transform(self.Xi), self.yi)
  File "/home/iaroslav/scikit-optimize/skopt/learning/gaussian_process/gpr.py", line 190, in fit
    super(GaussianProcessRegressor, self).fit(X, y)
  File "/usr/local/lib/python2.7/dist-packages/sklearn/gaussian_process/gpr.py", line 231, in fit
    self.L_ = cholesky(K, lower=True)  # Line 2
  File "/usr/local/lib/python2.7/dist-packages/scipy/linalg/decomp_cholesky.py", line 81, in cholesky
    check_finite=check_finite)
  File "/usr/local/lib/python2.7/dist-packages/scipy/linalg/decomp_cholesky.py", line 30, in _cholesky
    raise LinAlgError("%d-th leading minor not positive definite" % info)
numpy.linalg.linalg.LinAlgError: 31-th leading minor not positive definite
```
So here indeed the algorithm finds a point very close to global minimum (at which objective is -1.0), and then suggests the same (or very close) point multiple times which leads to exception.
One way to fix this would be to define maximum  number of tries for the algorithm to suggest a new point to evaluate, and if algorithm fails to do so, either rise a warning that it seems that the algorithm converged to global minimum, or / and select the next point to evaluate by randomly sampling from the parameter space.

**Comment by betatim** [comment]
I think adding an option to have the minimizer stop suggesting new points/decide it has converged would be the way to go here.

**Comment by iaroslav-ai** [comment]
Solved in recent version of skopt - I cannot reproduce this  error anymore, thus close the issue.

**Comment by iaroslav-ai** [comment]
Nope still having issues with this from time to time

**Comment by iaroslav-ai** [comment]
Hm not sure whether this should be solved within skopt or within sklearn

**Comment by betatim** [comment]
Random thought: why do we not get a warning that this point has been evaluated before? Too far away for that warning to be triggered?

What do you think about using a callback that terminates the optimization process "early" when points are too close? We have a callback for that but it isn't used by default.

**Comment by iaroslav-ai** [comment]
Yeah the matrix inverse associated with using gp's is always a pain in my experience to deal with. 
That is an interesting question @betatim . I do not have any concrete ideas for why there are no warnings.

Yes, one could use a dedicated termination criterion. But it might also happen that the same point was generated by accident, and maybe if you re - run the next point generation process it will actually give a different point. Maybe trying multiple times and if you fail after sufficient number of tries to produce a new point you either terminate or generate a new point at random.

**Comment by iaroslav-ai** [comment]
Here is something interesting. Consider a code below:
```python
import numpy as np
#from sklearn.gaussian_process import GaussianProcessRegressor
from skopt.learning.gaussian_process.gpr import GaussianProcessRegressor

X = np.random.rand(8, 3)
y = np.random.rand(8)

X[:3,:] = 0.0
y[:3] = 1.0

model = GaussianProcessRegressor()
model.fit(X, y)
yp = model.predict(X)
```
While `sklearn`'s `GaussianProcessRegressor` is working fine, the one from `skopt` fails.

**Comment by iaroslav-ai** [comment]
So two possible solutions that I could imagine are:

1. Fix `GaussianProcessRegressor` so that it does not rise the exception like the `sklearn`'s one

2. Add code to `Optimizer` so that when a similar point is generated it is replaced with a random point

Before doing any solutions, it makes sense to look into how exactly the code fails on circleci (what if some special non - similar combination of points fails the fitting of GPR?). I will try to reproduce the failure later today.

**Comment by iaroslav-ai** [comment]
Hm cannot reproduce the failure right now on my machine. It seems quite reproducible on circle though; Do you know if there is something like numpy seed fixed by `jupyter nbconvert`?

**Comment by betatim** [comment]
> 1. Fix `GaussianProcessRegressor` so that it does not rise the exception like the `sklearn`'s one

So in `sklearn` the same thing happens but they swallow the exception or the exception doesn't occur?

**Comment by iaroslav-ai** [comment]
I am actually not sure what they do, but this does not happen with `sklearn` estimator. There is something in `skopt` extension it seems

**Comment by betatim** [comment]
That makes two of us :)

**Comment by betatim** [comment]
Fixed by #544
